src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -345,14 +349,44 @@ private IAIState deliver()
         boolean success = true;
         boolean extracted = false;
         final IItemHandler workerInventory = worker.getInventoryCitizen();
+        final List<ItemStack> itemsToDeliver = job.getTaskListWithSameDestination((IRequest<? extends Delivery>) currentTask).stream().map(r -> r.getRequest().getStack()).collect(Collectors.toList());
+        final Map<Item, Integer> itemsCount = new HashMap<>();
```

Comment:
I understand the reason behind the complexity, but I would simplify this grossly.
We mostly want to avoid them delivering things they picked up along the way. So we just map the List to a map of ItemStorage here (new ItemStorage(stack)).

And in the bottom we just check for a contains new ItemStorage(stack).

(p.s.: Item map key doesn't work, because of items with nbts etc)

Reply:
Ok, I'm gonna do that tomorrow, its 3 am for me